x/stakeibc/keeper/msg_server_redeem_stake.go

```diff
@@ -0,0 +1,101 @@
+
+import (
+	"context"
+	"strconv"
+
+	"github.com/Stride-Labs/stride/x/stakeibc/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	distributionTypes "github.com/cosmos/cosmos-sdk/x/distribution/types"
+	stakingTypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+func (k Keeper) RedeemStake(goCtx context.Context, msg *types.MsgRedeemStake) (*types.MsgRedeemStakeResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// vars
+	sender, err := sdk.AccAddressFromBech32(msg.Creator)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "address is invalid: %s", msg.Creator)
+	}
+	coinString := strconv.Itoa(int(msg.Amount)) + msg.Denom
+	inCoin, err := sdk.ParseCoinNormalized(coinString)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidCoins, "could not parse inCoin: %s", coinString)
+	}
+	hostZone, err := k.GetHostZoneFromLocalDenom(ctx, msg.Denom)
+	if err != nil {
+		return nil, err
+	}
+	delegationAccount := hostZone.GetDelegationAccount()
+	withdrawAccount := hostZone.GetWithdrawalAccount()
+	connectionId := hostZone.GetConnectionId()
+	
+	// Safety checks
+	// Redemption amount must be positive
+	if !inCoin.IsPositive() {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidCoins, "amount must be greater than 0. found: %s", msg.Amount)
+	}
+	// Denom is valid
+	// Should we register stAssets somewhere and add an additional check here?
+	if types.IsStAsset(msg.Denom) != true {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidCoins, "denom is not a valid stAsset. found: %s", msg.Denom)
+	}
+	// TODO: REMOVE THIS, TESTING PURPOSES ONLY
+	// #################################################################
+	// Mint coins and send them to senders account
+	// err = k.bankKeeper.MintCoins(ctx, types.ModuleName, sdk.NewCoins(inCoin))
+	// if err != nil {
+	// 	return nil, err
+	// }
+	// err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, sender, sdk.NewCoins(inCoin))
+	// if err != nil {
+	// 	return nil, err
+	// }
+	// #################################################################
+	// Creator owns at least "amount" stAssets
+	balance := k.bankKeeper.GetBalance(ctx, sender, msg.Denom)
+	if balance.IsLT(inCoin) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidCoins, "balance is lower than redemption amount. redemption amount: %s, balance %s: ", msg.Amount, balance.Amount)
+	}
+
+	// Escrow user's balance
+	err = k.bankKeeper.SendCoinsFromAccountToModule(ctx, sender, types.ModuleName, sdk.NewCoins(inCoin))
+	if err != nil {
+		k.Logger(ctx).Info("Failed to send sdk.NewCoins(inCoins) from account to module")
+		panic(err)
+	}
+
+	// calculate the redemption rate
+	// when redeeming tokens, multiply stAssets by the exchange rate (allStakedAssets / allStAssets)
+	// TODO(TEST-7): Update redemption_rate via ICQ
+	var rate sdk.Dec
+	rate = hostZone.LastRedemptionRate
+	if hostZone.RedemptionRate.LT(rate) {
```

Comment:
This is pulled from QS, I imagine they added it as a simple way to mitigate redemption rate manipulation. Needs more thought but pushing to TEST-7

Reply:
Resurfacing this since I don't see any reason we would want to use LastRedemptionRate and I rolled TEST-7 into TEST-50 before seeing this PR comment. 

What's the redemption rate manipulation attack vector we're preventing here? Let's discuss